flaml/autogen/agent/agent.py

```diff
@@ -30,22 +31,79 @@ def _remember(self, memory):
+        oai_message["role"] = "function" if message.get("role") == "function" else "assistant"
+        self._oai_conversations[recipient.name].append(oai_message)
+
         recipient.receive(message, self)
 
-    def _receive(self, message, sender):
-        """Receive a message from another agent."""
-        print("\n", "-" * 80, "\n", flush=True)
-        print(sender.name, "(to", f"{self.name}):", flush=True)
-        print(message, flush=True)
-        self._conversations[sender.name].append({"content": message, "role": "user"})
+    def _receive(self, message: Union[Dict, str], sender):
+        """Receive a message from another agent.
 
-    def receive(self, message, sender):
+        Args:
+            message (dict or str): message from the sender. If the type is dict, it may contain the following reserved fields (All fields are optional).
+                1. "content": content of the message, can be None.
+                2. "function_call": a dictionary containing the function name and arguments.
+                3. "role": role of the message, can be "assistant", "user", "function".
+                    This field is only needed to distinguish between "function" or "assistant"/"user".
+                4. "name": In most cases, this field is not needed. When the role is "function", this field is needed to indicate the function name.
+            sender: sender of an Agent instance.
+        """
+        message = self._message_to_dict(message)
+        # print the message received
+        print(sender.name, "(to", f"{self.name}):\n", flush=True)
+        if message.get("role") == "function":
+            func_print = f"***** Response from calling function \"{message['name']}\" *****"
+            print(func_print, flush=True)
+            print(message["content"], flush=True)
+            print("*" * len(func_print), flush=True)
+        else:
+            if message.get("content") is not None:
+                print(message["content"], flush=True)
+            if "function_call" in message:
+                func_print = f"***** Suggested function Call: {message['function_call'].get('name', '(No function name found)')} *****"
+                print(func_print, flush=True)
+                print(
+                    "Arguments: \n",
+                    message["function_call"].get("arguments", "(No arguments found)"),
+                    flush=True,
+                    sep="",
+                )
+                print("*" * len(func_print), flush=True)
+        print("\n", "-" * 80, flush=True, sep="")
+
+        # create openai message to be appended to the conversation
+        oai_message = {k: message[k] for k in ("content", "function_call", "name") if k in message}
+        oai_message["role"] = "function" if message.get("role") == "function" else "user"
+        self._oai_conversations[sender.name].append(oai_message)
```

Comment:
In general, the message received is not necessarily an oai message.
@qingyun-wu

Reply:
See comment above

Reply:
See comment below.